src/realm/sync/noinst/client_reset.cpp

```diff
@@ -353,111 +461,115 @@ void client_reset::transfer_group(const Transaction& group_src, Transaction& gro
         auto table_src = group_src.get_table(table_key);
         StringData table_name = table_src->get_name();
-        if (!table_name.begins_with("class"))
-            continue;
         logger.debug("Removing objects in '%1'", table_name);
         auto table_dst = group_dst.get_table(table_name);
-        if (auto pk_col = table_dst->get_primary_key_column()) {
-            std::vector<std::pair<Mixed, ObjKey>> objects_to_remove;
-            for (auto obj : *table_dst) {
-                auto pk = obj.get_any(pk_col);
-                if (!table_src->find_primary_key(pk)) {
-                    objects_to_remove.emplace_back(pk, obj.get_key());
-                }
-            }
-            for (auto& pair : objects_to_remove) {
-                logger.debug("  removing '%1'", pair.first);
-                table_dst->remove_object(pair.second);
+
+        auto pk_col = table_dst->get_primary_key_column();
+        REALM_ASSERT_DEBUG(pk_col); // sync realms always have a pk
+        std::vector<std::pair<Mixed, ObjKey>> objects_to_remove;
+        for (auto obj : *table_dst) {
+            auto pk = obj.get_any(pk_col);
+            if (!table_src->find_primary_key(pk)) {
+                objects_to_remove.emplace_back(pk, obj.get_key());
             }
         }
-        else {
-            std::vector<std::pair<GlobalKey, ObjKey>> objects_to_remove;
-            for (auto obj : *table_dst) {
-                auto oid = table_dst->get_object_id(obj.get_key());
-                auto key_src = table_src->get_objkey(oid);
-                if (oid.hi() == new_file_id || !key_src || !table_src->is_valid(key_src))
-                    objects_to_remove.emplace_back(oid, obj.get_key());
-            }
-            for (auto& pair : objects_to_remove) {
-                logger.debug("  removing '%1'", pair.first);
-                table_dst->remove_object(pair.second);
-            }
+        for (auto& pair : objects_to_remove) {
+            logger.debug("  removing '%1'", pair.first);
+            table_dst->remove_object(pair.second);
         }
     }
 
     // Add objects that are present in src but absent in dst.
     for (auto table_key : group_src.get_table_keys()) {
+        if (!group_src.table_is_public(table_key))
+            continue;
         auto table_src = group_src.get_table(table_key);
         StringData table_name = table_src->get_name();
-        if (!table_name.begins_with("class"))
-            continue;
         logger.debug("Adding objects in '%1'", table_name);
         auto table_dst = group_dst.get_table(table_name);
-        auto pk_col = table_src->get_primary_key_column();
 
         for (auto& obj : *table_src) {
-            auto oid = table_src->get_object_id(obj.get_key());
-            auto key_dst = table_dst->get_objkey(oid);
+            auto src_pk = obj.get_primary_key();
+            auto key_dst = table_dst->find_primary_key(src_pk);
             if (!key_dst || !table_dst->is_valid(key_dst)) {
```

Comment:
I think the `is_valid` part can be removed. If the find succeeds, the key is valid.